src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -482,7 +482,188 @@ class G1RefineCardClosure: public CardTableEntryClosure {
+          process_consecutive();
+        }
+        // Remember it. Process later.
+        _dirties.add(curr);
+
+        *curr = G1CardTable::claimed_card_val();
+      } else {
+        // It's claimed or deferred because of scan_rem_set(). We ignore them.
+        assert(card_val == G1CardTable::clean_card_val() ||
+               card_val & G1CardTable::claimed_card_val() ||
+               card_val & G1CardTable::deferred_card_val(),
+               "If it's not dirty, it must be clean, claimed, or deferred");;
+      }
+
+      --processed_card;
+    }
+
+    if (!_dirties.is_empty()) {
+      process_consecutive();
+    }
+  }
+
+public:
+  G1ScanDirtyCardsHRClosure(G1RemSetScanState* scan_state,
+                            HeapRegionChunkClaimer* chunk_claimer,
+                            G1ScanObjsDuringUpdateRSClosure* update_rs_cl) :
+    _g1h(G1CollectedHeap::heap()),
+    _ct(_g1h->card_table()),
+    _scan_state(scan_state),
+    _chunk_claimer(chunk_claimer),
+    _update_rs_cl(update_rs_cl),
+    _current_hr(NULL),
+    _scan_top(NULL),
+    _dirties(),
+    _dirty_card_num(0) {}
+
+  bool do_heap_region(HeapRegion* r) {
+    uint hrm_index = r->hrm_index();
+    HeapWord* scan_top = _scan_state->scan_top(hrm_index);
+    if (scan_top == NULL) {
+      // Either empty or in collection set.
+      _chunk_claimer->claim_region(hrm_index);
+      return false;
+    }
+    _update_rs_cl->set_region(r);
+    _current_hr = r;
+    assert(scan_top > r->bottom(), "Must be non-empty");
+    _scan_top = scan_top;
+
+    const uint chunks_per_region = HeapRegionChunkClaimer::ChunksPerRegion;
+    assert(HeapRegion::GrainWords % chunks_per_region == 0, "Chunk size must be multiple of word");
+    const size_t words_per_chunk = HeapRegion::GrainWords / chunks_per_region;
+    // Claim chunks backwards. Block offset table has an affinity for backward iteration.
+    uint chunk = chunks_per_region;
+    do {
+      --chunk;
+      HeapWord* chunk_start = r->bottom() + chunk * words_per_chunk;
+      // We have to claim this chunk even it's over scan top, so that other workers
+      // can quickly skip this region by checking is_region_claimed().
+      bool res = _chunk_claimer->claim_chunk(_chunk_claimer->get_chunk_index(hrm_index, chunk));
+      if (res == false) {
+        // Failed to claim this chunk. Another thread must have succeeded.
+        continue;
+      }
+      if (chunk_start >= scan_top) {
+        // Nothing to scan.
+        continue;
+      }
+      HeapWord* chunk_end = chunk_start + words_per_chunk;
+      scan_chunk(chunk_start, MIN2(chunk_end, scan_top));
+    } while (chunk > 0);
+
+    _current_hr = NULL;
+    _scan_top = NULL;
+    return false;
+  }
+
+  size_t dirty_card_num() const {
+    return _dirty_card_num;
+  }
+};
+
+void G1RemSet::update_rem_set(G1ParScanThreadState* pss, uint worker_i, HeapRegionChunkClaimer* chunk_claimer) {
```

Comment:
original patch has: assert(SafepointSynchronize::is_at_safepoint(), "Sanity");

Reply:
I think it's ok.